// speedux/lib/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.queryGlobalState = exports.useMiddleware = exports.useReducer = exports.store = exports.Provider = exports.useGenerator = exports.useAsync = exports.useDispatch = exports.useHandler = exports.useActions = exports.useGlobalState = exports.createGlobalState = void 0;

var _store = _interopRequireDefault(require("./store"));

exports.store = _store["default"];

var _create = _interopRequireDefault(require("./create"));

exports.createGlobalState = _create["default"];

var _useGlobalState = _interopRequireDefault(require("./use-global-state"));

exports.useGlobalState = _useGlobalState["default"];

var _useActions = _interopRequireDefault(require("./use-actions"));

exports.useActions = _useActions["default"];

var _useHandler = _interopRequireDefault(require("./use-handler"));

exports.useHandler = _useHandler["default"];

var _useDispatch = _interopRequireDefault(require("./use-dispatch"));

exports.useDispatch = _useDispatch["default"];

var _useAsync = _interopRequireDefault(require("./use-async"));

exports.useAsync = _useAsync["default"];
exports.useGenerator = _useAsync["default"];

var _provider = _interopRequireDefault(require("./provider"));

exports.Provider = _provider["default"];

/**
 * Local imports
 */

/**
 * Re-exports
 */
// will be deprecated in the next major version
var useReducer = function useReducer() {
  _store["default"].useReducer.apply(_store["default"], arguments);

  _store["default"].update();
};

exports.useReducer = useReducer;

var useMiddleware = function useMiddleware() {
  return _store["default"].useMiddleware.apply(_store["default"], arguments);
};

exports.useMiddleware = useMiddleware;

var queryGlobalState = function queryGlobalState(query) {
  return _store["default"].getState(query);
};

exports.queryGlobalState = queryGlobalState;

_store["default"].create();
